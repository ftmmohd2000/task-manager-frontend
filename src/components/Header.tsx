import React from "react";
import { connect, ConnectedProps } from "react-redux";
import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { startDelete, startLogout } from "../actions/auth";

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, AnyAction>
) => ({
  logout: () => dispatch(startLogout()),
  deleteUser: () => dispatch(startDelete())
});

const connector = connect(undefined, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

type IHeaderProps = PropsFromRedux & {
  history: {
    push: (arg0: string) => void;
  };
};

export const Header = ({
  logout,
  deleteUser,
  history
}: IHeaderProps) => {
  const tryLogout = () => {
    logout();
    history.push("/");
  };

  const tryDeleteUser = () => {
    deleteUser();
    history.push("/");
  };

  return (
    <div>
      <h1>Task Manager</h1>
      <button id="logout" onClick={tryLogout}>
        Logout
      </button>
      <button id="deleteUser" onClick={tryDeleteUser}>
        Delete Account
      </button>
    </div>
  );
};

export default connector(Header);
