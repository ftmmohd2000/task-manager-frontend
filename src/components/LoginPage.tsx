import React, { useState } from "react";
import { connect, ConnectedProps } from "react-redux";
import { RouteComponentProps } from "react-router";
import { Link } from "react-router-dom";
import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { startLogin } from "../actions/auth";
import { startSetTasks } from "../actions/task";

const mapDispatchToProps = (dispatch: ThunkDispatch<{}, {}, AnyAction>) => ({
  setTasks: () => dispatch(startSetTasks()),
  login: (email: string, password: string) =>
    dispatch(startLogin(email, password))
});

const connector = connect(undefined, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = PropsFromRedux & RouteComponentProps;

const LoginPage = ({ history, setTasks, login }: Props) => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState("");

  const tryLogin = (e: any) => {
    e.preventDefault();
    login(email, password)
      .then(() => {
        setTasks();
        history.push("/dashboard");
      })
      .catch(() => {
        setError("Invalid Credentials");
      });
  };

  return (
    <div>
      <h1>Login Page</h1>
      <form onSubmit={tryLogin}>
        <label>email:</label>
        <input
          type="text"
          value={email}
          onChange={(e: any) => {
            setEmail(e.target.value);
          }}
        />
        <label>password:</label>
        <input
          type="password"
          value={password}
          onChange={(e: any) => {
            setPassword(e.target.value);
          }}
        />
        <button type="submit">Login</button>
      </form>
      {error && <p>{error}</p>}
      <Link to="/signup">Signup</Link>
    </div>
  );
};

export default connector(LoginPage);
