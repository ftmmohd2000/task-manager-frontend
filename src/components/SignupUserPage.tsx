import React, { useState } from "react";
import { RouteComponentProps } from "react-router";
import { startSignup } from "../actions/auth";
import { connect } from "react-redux";
import { ThunkDispatch } from "redux-thunk";
import { AnyAction } from "redux";

interface ISignupUserPageProps extends RouteComponentProps {
  signup: (
    arg0: string,
    arg1: string,
    arg2: string,
    arg3: number
  ) => Promise<any>;
}
export const SignupUserPage = (
  props: ISignupUserPageProps
) => {
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState("");
  const [age, setAge] = useState(0);

  const tryLogin = (e: any) => {
    e.preventDefault();
    props
      .signup(name, email, password, age)
      .then(() => {
        props.history.push("/dashboard");
      })
      .catch(() => {
        setError("Invalid Credentials");
      });
  };

  return (
    <div>
      <h1>Signup Page</h1>
      <form onSubmit={tryLogin}>
        <p>email: </p>
        <input
          type="text"
          value={email}
          onChange={(e: any) => {
            setEmail(e.target.value);
          }}
        />
        <p>password: </p>
        <input
          type="password"
          value={password}
          onChange={(e: any) => {
            setPassword(e.target.value);
          }}
        />
        <p>name: </p>
        <input
          type="text"
          value={name}
          onChange={(e: any) => {
            setName(e.target.value);
          }}
        />
        <p>age: </p>
        <input
          type="number"
          value={age}
          onChange={(e: any) => {
            setAge(e.target.value);
          }}
        />
        <button>SignUp</button>
      </form>
      {error && <p>{error}</p>}
    </div>
  );
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, AnyAction>
) => ({
  signup: (
    name: string,
    email: string,
    password: string,
    age: number = 0
  ) => dispatch(startSignup(name, email, password, age))
});
export default connect(
  undefined,
  mapDispatchToProps
)(SignupUserPage);
