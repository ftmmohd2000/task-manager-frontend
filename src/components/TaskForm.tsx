import { History } from "history";
import React, { useState } from "react";
import TaskType from "../types/task";

interface ITaskFormProps {
  history: History<any>;
  saveTask: (arg0: string, arg1: boolean) => Promise<void>;
  task?: TaskType;
}

export const TaskForm = ({
  history,
  saveTask,
  task
}: ITaskFormProps) => {
  const [description, setDescription] = useState(
    task?.description || ""
  );
  const [completed, setCompleted] = useState(
    task?.completed || false
  );

  const handleFormSubmit = (e: any) => {
    e.preventDefault();
    saveTask(description, completed);
    history.push("/dashboard");
  };

  return (
    <form onSubmit={handleFormSubmit}>
      <input
        type="text"
        value={description}
        onChange={(e: any) =>
          setDescription(e.target.value)
        }
      />
      <input
        type="checkbox"
        checked={completed}
        onChange={(e: any) => setCompleted(!completed)}
      />
      <button type="submit">Add Task</button>
    </form>
  );
};

export default TaskForm;
