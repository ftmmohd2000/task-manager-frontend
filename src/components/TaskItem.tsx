import React from "react";
import TaskType from "../types/task";
import { Link } from "react-router-dom";

interface ITaskItemProps {
  task: TaskType;
  handleRemove: (arg0: string) => void;
}

export const TaskItem = ({
  handleRemove,
  task: { description, completed, _id }
}: ITaskItemProps) => {
  const removeItem = (e: any) => {
    handleRemove(_id);
  };
  return (
    <div>
      <p>{description}</p>
      <p>Completed: {completed ? "yes" : "no"}</p>
      <button onClick={removeItem}>Remove</button>
      <Link to={`/edit/${_id}`}>Edit</Link>
    </div>
  );
};

export default TaskItem;
