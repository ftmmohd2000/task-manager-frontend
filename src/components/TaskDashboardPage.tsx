import React from "react";
import { RouteComponentProps } from "react-router";
import { connect, ConnectedProps } from "react-redux";
import { Link } from "react-router-dom";
import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { startRemoveTask } from "../actions/task";
import Header from "./Header";
import TaskList from "./TaskList";
import TaskType from "../types/task";

const mapStateToProps = (state: any) => ({
  tasks: state.task.tasks
});

const mapDispatchToProps = (dispatch: ThunkDispatch<{}, {}, AnyAction>) => ({
  removeTask: (id: string) => dispatch(startRemoveTask(id))
});

const connector = connect(mapStateToProps, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = RouteComponentProps &
  PropsFromRedux & {
    tasks: TaskType[];
  };

const TaskDashboardPage = ({ removeTask, history, tasks }: Props) => {
  const handleRemove = (id: string) => {
    removeTask(id);
  };

  return (
    <div>
      <Header history={history} />
      <TaskList tasks={tasks} handleRemove={handleRemove} />
      <Link to="/add">Add Task</Link>
    </div>
  );
};

export default connector(TaskDashboardPage);
