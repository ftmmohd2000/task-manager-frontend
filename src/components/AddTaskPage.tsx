import React from "react";
import TaskForm from "./TaskForm";
import { RouteComponentProps } from "react-router";
import { ConnectedProps, connect } from "react-redux";
import { ThunkDispatch } from "redux-thunk";
import { AnyAction } from "redux";
import { startAddTask } from "../actions/task";

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, AnyAction>
) => ({
  addTask: (description: string, completed: boolean) =>
    dispatch(startAddTask(description, completed))
});

const connector = connect(undefined, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

type IAddTaskPageProps = RouteComponentProps &
  PropsFromRedux;

export const AddTaskPage = ({
  history,
  addTask
}: IAddTaskPageProps) => {
  return (
    <div>
      <TaskForm history={history} saveTask={addTask} />
    </div>
  );
};

export default connector(AddTaskPage);
