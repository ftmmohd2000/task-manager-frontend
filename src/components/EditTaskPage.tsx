import React from "react";
import { connect, ConnectedProps } from "react-redux";
import {
  Redirect,
  RouteComponentProps
} from "react-router";
import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { startEditTask } from "../actions/task";
import { ReduxStateType } from "../store/configureStore";
import TaskType from "../types/task";
import TaskForm from "./TaskForm";

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, AnyAction>
) => ({
  editTask: (id: string, updates: Partial<TaskType>) =>
    dispatch(startEditTask(id, updates))
});

const mapStateToProps = (
  state: ReduxStateType,
  props: RouteComponentProps<{ id: string }>
) => ({
  task: state.task.tasks.find(
    (task) => task._id === props.match.params.id
  ),
  id: props.match.params.id
});

const connector = connect(
  mapStateToProps,
  mapDispatchToProps
);

type PropsFromRedux = ConnectedProps<typeof connector>;

type IEditTaskPageProps = RouteComponentProps<{
  id: string;
}> &
  PropsFromRedux;

export const EditTaskPage = ({
  task,
  history,
  id,
  editTask
}: IEditTaskPageProps) => {
  const startEditTask = (
    description: string,
    completed: boolean
  ) => {
    return editTask(id, {
      description,
      completed
    });
  };

  return task ? (
    <div>
      <TaskForm
        history={history}
        task={task}
        saveTask={startEditTask}
      />
    </div>
  ) : (
    <Redirect to="/dashboard" />
  );
};

export default connector(EditTaskPage);
