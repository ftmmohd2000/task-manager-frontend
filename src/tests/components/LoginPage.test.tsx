import React from "react";
import { shallow } from "enzyme";
import { LoginPage } from "../../components/LoginPage";

let wrapper: any,
  setTasks: () => Promise<void>,
  login: (arg0: string, arg1: string) => Promise<void>;

beforeEach(() => {
  const routeComponentPropsMock = {
    history: { push: jest.fn() } as any,
    match: {} as any,
    location: {} as any
  };
  login = jest.fn(() => Promise.resolve());
  setTasks = jest.fn();
  wrapper = shallow(
    <LoginPage
      setTasks={setTasks}
      login={login}
      {...routeComponentPropsMock}
    />
  );
});

test("should render LoginPage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should run login function when button is clicked", () => {
  wrapper.find("input").at(0).value = "someEmail";
  wrapper.find("input").at(1).value = "somePassword";
  wrapper.find("form").simulate("submit", {
    preventDefault: () => {}
  });
  expect(login).toHaveBeenCalled();
});
