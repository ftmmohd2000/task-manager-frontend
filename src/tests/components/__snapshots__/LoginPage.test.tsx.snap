// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render LoginPage correctly 1`] = `
<div>
  <h1>
    Login Page
  </h1>
  <form
    onSubmit={[Function]}
  >
    <label>
      email:
    </label>
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
    <label>
      password:
    </label>
    <input
      onChange={[Function]}
      type="password"
      value=""
    />
    <button
      type="submit"
    >
      Login
    </button>
  </form>
  <Link
    to="/signup"
  >
    Signup
  </Link>
</div>
`;
